test/client/dio_interceptor_test.dart

```diff
@@ -0,0 +1,231 @@
+      clientId: 'test_client_id',
+      clientSecret: 'test_client_secret',
+      headerName: 'X-Signature',
+      signaturePattern: '{ClientId}:{Nonce}:{Timestamp}:{SignatureBody}',
+      disableAutoRetryOnClockSkew: false,
+      clockSkew: Duration(milliseconds: clockskewMS),
+    );
+
+    // Create a Dio instance with mock adapter
+    final dio = Dio(BaseOptions(
+      validateStatus: (status) => true,
+    ));
+
+    final dioAdapter = DioAdapter(dio: dio);
+    dio.httpClientAdapter = dioAdapter;
+
+    final serverDate = DateTime.now().toUtc().add(Duration(milliseconds: 6000));
+    final serverDateIsoString = serverDate.toIso8601String();
+
+    // Create the interceptor
+    final interceptor = RequestsSignatureInterceptor(
+      options,
+      dio,
+      getTime: (request) {
+        return DateTime.now().toUtc().millisecondsSinceEpoch ~/ 1000;
+      },
+      getDateHeader: () {
+        return serverDateIsoString;
+      },
+    );
+
+    // Add the interceptor to Dio
+    dio.interceptors.add(interceptor);
+
+    // Register the mock response
+    dioAdapter.onGet('https://google.ca/', (request) {
+      // Check if the server returns its date time header
+      final hasServerDateHeader = serverDateIsoString.isNotEmpty;
+
+      // Respond with unauthorized or ok based on clock skew and tolerance settings
+      if (options.disableAutoRetryOnClockSkew || toleranceMS < clockskewMS) {
+        request.reply(HttpStatus.unauthorized, {});
+      } else if (hasServerDateHeader) {
+        request.reply(HttpStatus.ok, {});
+      } else {
+        request.reply(HttpStatus.unauthorized, {});
+      }
+    });
+
+    // Act
+    final response = await dio.get('https://google.ca/');
+
+    // Assert
+    expect(response.statusCode, HttpStatus.unauthorized);
+  });
+
+  test('Auto-retry disabled when clock skew option is disabled', () async {
+    final options = RequestsSignatureOptions(
+      clientId: 'test_client_id',
+      clientSecret: 'test_client_secret',
+      headerName: 'X-Signature',
+      signaturePattern: '{ClientId}:{Nonce}:{Timestamp}:{SignatureBody}',
+      disableAutoRetryOnClockSkew: true,
+    );
+
+    final dio = Dio();
+    final dioAdapter = DioAdapter(dio: dio);
+    dio.httpClientAdapter = dioAdapter;
+
+    final serverDate = DateTime.now().toUtc().add(Duration(milliseconds: 6000));
+    final serverDateIsoString = serverDate.toIso8601String();
+
+    // Create the interceptor
+    final interceptor = RequestsSignatureInterceptor(
+      options,
+      dio,
+      getTime: (request) {
+        return DateTime.now().toUtc().millisecondsSinceEpoch ~/ 1000;
+      },
+      getDateHeader: () {
+        return serverDateIsoString;
+      },
+    );
+
+    // Add the interceptor to Dio
+    dio.interceptors.add(interceptor);
+
+    int requestCount = 0;
+
+    dioAdapter.onGet('https://google.ca/', (request) {
+      requestCount++;
+      request.reply(HttpStatus.ok, {});
```

Comment:
I think the behavior of the mock server should be the same in all tests so that you have a uniform behavior. You should likely extract it and reuse it in all tests.

Consequently, this test is bad because it doesn't validate the correct behavior: If the service always reply `ok`, the interceptor never needs to retry and the clock skew behavior is completely skipped.